### PR TITLE
Replace stdout debug by log in pattern matcher

### PR DESCRIPTION
**Pull request by ngeiswei**
This is to address issue #147 . All stdout debug messages have been replaced by fine debugging. The reason I use fine is because most pattern matcher tests enable debug level and this would slow down considerably run-time pattern matching. Generally I think we can assume that log levels behave that way

1. INFO: contain the bare minimum
2. DEBUG: contain useful information for debugging but do not slow down run-time substantially
3. FINE: may slow down run-time substantially and may not be used in production

The difference in performance is at DEBUG level is not significant, see statistics obtained by running scripts/query/benchmark_query.sh (which runs 20 times all pattern matcher unit tests, and spits out stats).

```
Before replacing stdout by log
========================
N   min    max    sum     mean     stddev
20  13.57  17.99  309.61  15.4805  1.43348
```

```
After replacing stdout by log
=======================
N   min    max    sum    mean    stddev
20  13.77  17.86  313.1  15.655  1.52332
```

The new version is 1% slower, but I don't think it is statistically significant given the standard deviation.

I think it's ready to be merged, but @linas I'll probably let you merge just in case.

### CODE REVIEW

File: opencog/query/InitiateSearchCB.cc [hunk 1]
```diff
@@ -565,14 +558,12 @@ bool InitiateSearchCB::link_type_search(PatternMatchEngine *pme)
...
-	size_t i = 0;
-#endif
+	size_t i = 0, hsz = handle_set.size();
 	for (const Handle& h : handle_set)
 	{
-		dbgprt("yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy\n");
-		dbgprt("Loop candidate (%lu/%lu):\n%s\n", ++i, handle_set.size(),
-		       h->toShortString().c_str());
+		LAZY_LOG_FINE << "yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy\n"
+		              << "Loop candidate (" << i+1 << "/" << hsz << "):\n"
```
Comment (linas) [inline opencog/query/InitiateSearchCB.cc]: You've introduced a bug here; i no longer increments.  I will fix this shortly n a different pull request.

Comment (ngeiswei) [inline opencog/query/InitiateSearchCB.cc]: Oh sorry about that!

File: opencog/query/InitiateSearchCB.cc [hunk 2]
```diff
@@ -667,16 +661,14 @@ bool InitiateSearchCB::variable_search(PatternMatchEngine *pme)
...
-	size_t i = 0;
-#endif
+	size_t i = 0, hsz = handle_set.size();
 	for (const Handle& h : handle_set)
 	{
-		dbgprt("zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz\n");
-		dbgprt("Loop candidate (%lu/%lu):\n%s\n", ++i, handle_set.size(),
-		       h->toShortString().c_str());
+		LAZY_LOG_FINE << "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz\n"
+		              << "Loop candidate (" << i+1 << "/" << hsz << "):\n"
```
Comment (ngeiswei) [inline opencog/query/InitiateSearchCB.cc]: Same bug here.

**Comment by linas** [comment]
I dunno how much you use this, but I am getting to really hate this change. It makes debugging extra hard: there is much more work to do, much more typing, the logs are much hard to read.   This simply ws not a good idea.

This needs to be redesigned so that (a) it is easy to use,  (b) the log messages are readable.   The old way of doing things satisfied (a) and (b), the new way does not.

This is now the 5th or 6th time I have cursed this change, and I have Io curse it a few more time, I plan to revert it.

**Comment by ngeiswei** [comment]
Much more work to do!!! It is so much easier! No need to recompile, just add a line of scheme code where you want to use it. I don't see how it's harder to use.

**Comment by linas** [comment]
Its like 5 lines of scheme code, impossible to memorize, so you have to
either copy it somewhere and try to remember where you copied it, or you
need to reconstruct it from sratch which means using ,apropos, figuring out
which calls you need to make to set it up, and how to use them, and tehn
make several mistakes before you can actually use it.

Oh, and to add insult to injury, if you recompile the code, you have to do
this again .. and again and again, dozens of times a day.  cut-n-paste the
same boilerplate over and over. Its a major hassle.

With a compile time directive, its easy to find, you enable it once in the
morning you never have to touch it again, until you are done.   It would be
great if it was that easy, but its just not.

Oh, and then also everything else in the cogserver that is set to log fine
spoils the output, so you have to go in by hand, and change all those OTHER
users of fine(), and disable them, and recompile.   The logger is
absolutely horrific in terms of ease-of-use.

**Pull request merged**